### ISSUE TITLE
data loss (unknown cause)

### ISSUE THREAD

**Issue by aurisnoctis** [mpcjanssen/simpletask-android#954]
Hello.

- [x] checked for following filters to avoid duplicates: `is:issue is:open "data loss"`; relevant issues like https://github.com/mpcjanssen/simpletask-android/issues/205 were fixed

### What happened

- were looking for something I had saved with tag `@notes` in Simpletask
- used search inside app with known text fragment
- saw respective task flashing briefly on the screen during the live search for it, so I was pretty sure it was what I was looking for
- then the task disappeared from the live search list of tasks
- explicitly made sure no filters or searches were active and typed same search phrase again, which yielded no tasks anymore
- searched for several variations and text fragments that were part of the searched-for task, but no corresponding tasks are found
- again removed all filters / searches and checked from left menu just `@notes`: it contained only 10 items of once at least 80 as I know from a version of todo.txt from May 2018; I took a lot of notes over the years that I never marked complete; overall I now have 700 tasks and not so long ago had more than 1300. Uh-oh.

### What I tried to recover data

- check Menu > History, as as explained to me before in #286 , and sent all those available versions to my computer to compare as text files. They don't go back far enough to see at what point in time all those tasks were lost. I noticed the data loss the day before yesterday, but it must have happened earlier (2 days before at least); also Debugging info doesn't go back far enough
- check my own independent online backup, but that was broken due to other reasons

### Things I suspect could play a role

- `Menu > Settings > Features > Edit shared tasks` is active even though issue #721 is still open (?)
- somehow binary symbols made it into the file - could they disturb the processing or were caused by a disturbed process in Simpletask? In vim the symbols were shown as `^@`. One was written in a place where the task was still there, but text was clearly missing; one appeared at the end of a task line. The binary symbols in the file caused for instance for the text file to be recognized as a binary file:
```
> file todo.txt
todo.txt_copy: data
> dos2unix todo.txt
dos2unix: Binary symbol 0x00 found at line 50
```
- very speculative: `Menu > Settings > Features > Keep selection` is active; it sometimes seems to cause inconsistencies (I haven't files as issues yet) when a task is marked complete but is still selected and the list should filter out completed tasks, but the task is still shown because it is still selected
- `Battery optimization` was switched on on my phone for all apps except Google Play Services; I now included Simpletask in the list of not-optimized apps

### What I expect

- not a miracle, rather to document it for future reference
- maybe some ideas, for instance if #721 is still relevant or if anybody else had a similar problem
- if 1300 tasks is too much a burden for Simpletask

Thanks!

**Comment by mpcjanssen** [comment]
Nothing obviously seems flaky about this. Are you using Simpletask cloudless?

**Comment by aurisnoctis** [comment]
Yes.

**Comment by mpcjanssen** [comment]
Do you synchronize the todo.txt file with some other software in the background?

**Comment by aurisnoctis** [comment]
Yes, I have an app on my phone that connects to a cloud based on ownCloud, hosted by a large scientific organization. However, the mobile app doesn't sync regularly by itself. From I deduce:

1. Oh no, my todo.txt wasn't saved regularly, only when I pressed the sync button, which wasn't often enough to trace when and why half of my tasks have gone missing.
2. Probably the ownCloud client app isn't the culprit as it was "sleeping" all this time never springing into action. Basically I just happened to have my files in a folder that was in a list of folders to be synced when the sync button is pressed.

**Comment by mpcjanssen** [comment]
@aurisnoctis background synchronisation may trigger Simpletask to re-read the todo.txt file. I can imagine that re-reading the file at the wrong time might lose data.

**Comment by mpcjanssen** [comment]
A 5 second wait is definitely not right (especially with cloudless). However Simpletask tries its best to not lose data or end up in an inconsistent state. As a result saving the task list has become progressively slower. (Things like Lua filters don't help)

Tip for quick backup, use the Android share function to mail yourself the current list.

**Comment by scheunengeist** [comment]
@aurisnoctis Just as a sidenote, I have an S4 mini as well, running LineageOS. Saving a task (ST cloudless from f-droid) has indeed become a bit slower, but it doesn't take more than appr. one second here. Currently about 260 tasks in the list.
Of course, none of this is any help concerning the data loss issue... :)

**Comment by aurisnoctis** [comment]
@scheunengeist Thanks - I yet have to switch to LineageOS, still running CyanogenMod :[]. For me, it is always around 5 seconds, whenever the overview list is refreshed, e.g., after saving a new task.

@mpcjanssen OK, thanks for the link. First, I will migrate all of my todo-related files out of any backup folders, to be sure there are no side effects from non-ST software. The next step would be the newest version from your website. Currently, I am running ST version 10.2.9. That's also the most current one on Google Play for me.

Is it guaranteed that any update of the app lets a current process of re-writing the todo file finish?

Thanks!

**Comment by mpcjanssen** [comment]
@aurisnoctis I don't think updating an app guarantees anything. But I would expect saving preferences is atomic.  And that's where  ST saves 
the Todo list internally It might be a good idea to disable automatic updates though.

Also the version from the link I posted is the current development version which has some changes to making file saves and UI updates faster so it might be worth checking out.
